File: src-tauri/src/discord.rs
//! Discord Rich Presence Integration
//! Shows currently playing track in Discord

use discord_rich_presence::{activity, DiscordIpc, DiscordIpcClient};
use parking_lot::Mutex;
use std::sync::Arc;

// Discord Application ID (you'd create this in Discord Developer Portal)
const DISCORD_APP_ID: &str = "1234567890123456789"; // Replace with actual ID

lazy_static::lazy_static! {
    static ref DISCORD_CLIENT: Arc<Mutex<Option<DiscordIpcClient>>> = Arc::new(Mutex::new(None));
}

/// Connect to Discord
#[tauri::command]
pub async fn discord_connect() -> Result<(), String> {
    let mut client = DiscordIpcClient::new(DISCORD_APP_ID)
        .map_err(|e| format!("Failed to create Discord client: {}", e))?;

    client
        .connect()
        .map_err(|e| format!("Failed to connect to Discord: {}", e))?;

    let mut guard = DISCORD_CLIENT.lock();
    *guard = Some(client);

    Ok(())
}

/// Disconnect from Discord
#[tauri::command]
pub async fn discord_disconnect() -> Result<(), String> {
    let mut guard = DISCORD_CLIENT.lock();
    if let Some(mut client) = guard.take() {
        client
            .close()
            .map_err(|e| format!("Failed to disconnect from Discord: {}", e))?;
    }
    Ok(())
}

/// Set Discord activity (rich presence)
#[tauri::command]
pub async fn discord_set_activity(
    details: String,
    state: String,
    large_image_key: String,
    large_image_text: String,
    small_image_key: Option<String>,
    small_image_text: Option<String>,
    start_timestamp: Option<i64>,
    end_timestamp: Option<i64>,
) -> Result<(), String> {
    let mut guard = DISCORD_CLIENT.lock();
    let client = guard
        .as_mut()
        .ok_or_else(|| "Discord not connected".to_string())?;

    let mut activity_builder = activity::Activity::new()
        .details(&details)
        .state(&state)
        .assets(
            activity::Assets::new()
                .large_image(&large_image_key)
                .large_text(&large_image_text),
        );

    // Add small image if provided
    if let (Some(key), Some(text)) = (&small_image_key, &small_image_text) {
        activity_builder = activity_builder.assets(
            activity::Assets::new()
                .large_image(&large_image_key)
                .large_text(&large_image_text)
                .small_image(key)
                .small_text(text),
        );
    }

    // Add timestamps if provided
    if let Some(start) = start_timestamp {
        let mut timestamps = activity::Timestamps::new().start(start);
        if let Some(end) = end_timestamp {
            timestamps = timestamps.end(end);
        }
        activity_builder = activity_builder.timestamps(timestamps);
    }

    // Add buttons
    activity_builder = activity_builder.buttons(vec![activity::Button::new(
        "Listen on Apple Music",
        "https://music.apple.com",
    )]);

    client
        .set_activity(activity_builder)
        .map_err(|e| format!("Failed to set activity: {}", e))?;

    Ok(())
}

/// Clear Discord activity
#[tauri::command]
pub async fn discord_clear_activity() -> Result<(), String> {
    let mut guard = DISCORD_CLIENT.lock();
    let client = guard
        .as_mut()
        .ok_or_else(|| "Discord not connected".to_string())?;

    client
        .clear_activity()
        .map_err(|e| format!("Failed to clear activity: {}", e))?;

    Ok(())
}

File: electron/main/discord.ts
// Discord Rich Presence Integration
// Uses discord-rpc npm package

let client: any = null;
let isConnected = false;

// Discord Application ID - Create your own at https://discord.com/developers/applications
const DISCORD_APP_ID = '0000000000000000000';

export async function discordConnect(): Promise<void> {
  try {
    // Dynamic import to avoid crash if discord-rpc not available
    const { Client } = await import('discord-rpc');
    client = new Client({ transport: 'ipc' });
    await client.login({ clientId: DISCORD_APP_ID });
    isConnected = true;
    console.log('[TUFFAHI] Connected to Discord');
  } catch (error) {
    isConnected = false;
    throw new Error(`Failed to connect to Discord: ${error}`);
  }
}

export async function discordDisconnect(): Promise<void> {
  if (!client || !isConnected) return;

  try {
    await client.destroy();
    isConnected = false;
    client = null;
    console.log('[TUFFAHI] Disconnected from Discord');
  } catch (error) {
    throw new Error(`Failed to disconnect from Discord: ${error}`);
  }
}

export async function discordSetActivity(params: {
  details: string;
  state: string;
  largeImageKey: string;
  largeImageText: string;
  smallImageKey?: string;
  smallImageText?: string;
  startTimestamp?: number;
  endTimestamp?: number;
}): Promise<void> {
  if (!client || !isConnected) {
    throw new Error('Discord not connected');
  }

  const activity: Record<string, any> = {
    details: params.details,
    state: params.state,
    largeImageKey: params.largeImageKey,
    largeImageText: params.largeImageText,
  };

  if (params.smallImageKey) {
    activity.smallImageKey = params.smallImageKey;
    activity.smallImageText = params.smallImageText;
  }

  if (params.startTimestamp) {
    activity.startTimestamp = params.startTimestamp;
    if (params.endTimestamp) {
      activity.endTimestamp = params.endTimestamp;
    }
  }

  await client.setActivity(activity);
}

export async function discordClearActivity(): Promise<void> {
  if (!client || !isConnected) {
    throw new Error('Discord not connected');
  }

  await client.clearActivity();
}
